Feature | Implement SqlConnection.GetSchemaAsync

edwardneal opened the pull request:
Contributes to #646, adding `SqlConnection.GetSchemaAsync`.

I can't add the IDbColumnSchemaGenerator method because the .NET Framework version of SqlDataReader doesn't yet implement that interface. #2967 tracks this. `SqlDataReader.GetSchemaTableAsync` requires some extra work with the state machine to guarantee that the column metadata is available, and I'll submit a separate PR for this.

I've added some basic documentation, but this effectively follows the `SqlCommand.[Something]Async` pattern of "An asynchronous version of [Something], which <etc.>"

I've also moved the netfx GetSchema methods from SqlConnectionHelper.cs to SqlConnection.cs, to align that part of it with netcore. This contributes to the merge slightly.

edwardneal committed:
Implemented asynchronous schema APIs

edwardneal committed:
Remove erroneously-added reference to shared csproj

roji commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/ProviderBase/DbMetaDataFactory.cs:
```diff
@@ -175,13 +177,18 @@ private DataTable ExecuteCommand(DataRow requestedCollectionRow, string[] restri
+                cancellationToken.ThrowIfCancellationRequested();
+#if NET
+                DataTable schemaTable = isAsync ? await reader.GetSchemaTableAsync(cancellationToken).ConfigureAwait(false) : reader.GetSchemaTable();
+#else
```
Would it make sense to have the .NET Framework path also use an async API here? Unless I'm missing something, the exact same methods/code could be define for both, with the sole extension that in .NET Framework it wouldn't override?

edwardneal replied:
I can do that; I thought it might look a little clumsy for the .NET Framework version of Microsoft.Data.SqlClient to have a dedicated async method which isn't present on System.Data.SqlClient and which doesn't actually run asynchronously.

(GetSchemaTable just constructs a DataTable from the cached metadata - there's nothing which we can make async.)

edwardneal replied:
Running this through CI now...

The IDbColumnSchemaGenerator method will need to be done after (or possibly as part of) #2967.

roji replied:
> GetSchemaTable just constructs a DataTable from the cached metadata - there's nothing which we can make async

Oh I must be misunderstanding what this is about then - I thought the idea here is to actually get the schema information (asynchronously).

In any case, the idea is to have the least amount of separate code paths here for .NET Framework vs. modern .NET...

edwardneal replied:
I implemented this, but on re-checking, we access a property within SqlDataReader which performs a read from the network. I'm going to revert the SqlDataReader changes tomorrow so that I don't introduce async-over-sync within the new API. I'll submit the new SqlDataReader API in a subsequent PR, which will need more review since it'll involve touching the state machine.

SqlConnection.GetSchemaAsync is much simpler, since it just layers atop SqlCommand. This PR can still introduce that safely.

roji commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/ProviderBase/DbMetaDataFactory.cs:
```diff
@@ -107,7 +109,7 @@ protected virtual void Dispose(bool disposing)
         }
 
-        private DataTable ExecuteCommand(DataRow requestedCollectionRow, string[] restrictions, DbConnection connection)
+        private async ValueTask<DataTable> ExecuteCommandAsync(DataRow requestedCollectionRow, string[] restrictions, DbConnection connection, bool isAsync, CancellationToken cancellationToken)
```
Really nice to be seeing proper async/await in SqlClient!

edwardneal committed:
Implemented SqlDataReader.GetSchemaTableAsync

edwardneal committed:
ref project cleanup

edwardneal committed:
Add default parameter value to ref projects

edwardneal committed:
Merge branch 'main' into issue-646

edwardneal committed:
Revert addition of SqlDataReader API

This introduces async-over-sync, removing such requires work with the state machine and separate review.

paulmedynski commented:
@edwardneal - Can you resolve the conflicts, and then I will kick off another CI run?

edwardneal committed:
Merge main

edwardneal committed:
Account for SqlDataReader merge and .NET Standard support

edwardneal commented:
Thanks @paulmedynski, I've just brought the branch up to date.

paulmedynski commented:
/azp run

paulmedynski reviewed (changes requested):
A few inconsistencies and questions.

edwardneal committed:
Response to code review

Remove SqlDataReader.GetSchemaTableAsync (and references to such.)
Expand test coverage of both GetSchema and GetSchemaAsync.
Correct typo in documentation for GetSchema and GetSchemaAsync.

edwardneal commented on doc/snippets/Microsoft.Data.SqlClient/SqlConnection.xml:
```diff
@@ -1363,6 +1363,22 @@ For more information on working with events, see [Connection Events](https://lea
+        The cancellation token.
+      </param>
+      <summary>
+        An asynchronous version of <see cref="M:Microsoft.Data.SqlClient.SqlConnection.GetSchema()" />, which returns schema information for the data source of this <see cref="T:Microsoft.Data.SqlClient.SqlConnection" />. For more information about scheme, see <see href="https://learn.microsoft.com/sql/connect/ado-net/sql-server-schema-collections">SQL Server Schema Collections</see>.
```
Thanks - corrected in GetSchemaAsync and GetSchema.

edwardneal commented on src/Microsoft.Data.SqlClient/netcore/ref/Microsoft.Data.SqlClient.cs:
```diff
@@ -993,6 +993,21 @@ public event Microsoft.Data.SqlClient.SqlInfoMessageEventHandler InfoMessage { a
         public override System.Data.DataTable GetSchema(string collectionName) { throw null; }
         /// <include file='../../../../doc/snippets/Microsoft.Data.SqlClient/SqlConnection.xml' path='docs/members[@name="SqlConnection"]/GetSchemaCollectionNameRestrictionValues/*'/>
         public override System.Data.DataTable GetSchema(string collectionName, string[] restrictionValues) { throw null; }
+#if NET
```
The file is also used to build the netstandard2.0 ref assemblies, which doesn't provide this member in the base class.

edwardneal commented on src/Microsoft.Data.SqlClient/netcore/ref/Microsoft.Data.SqlClient.cs:
```diff
@@ -993,6 +993,21 @@ public event Microsoft.Data.SqlClient.SqlInfoMessageEventHandler InfoMessage { a
+        public override System.Threading.Tasks.Task<System.Data.DataTable> GetSchemaAsync(string collectionName, System.Threading.CancellationToken cancellationToken = default) { throw null; }
+        /// <include file='../../../../doc/snippets/Microsoft.Data.SqlClient/SqlConnection.xml' path='docs/members[@name="SqlConnection"]/GetSchemaCollectionNameRestrictionValuesAsync/*'/>
+        public override System.Threading.Tasks.Task<System.Data.DataTable> GetSchemaAsync(string collectionName, string[] restrictionValues, System.Threading.CancellationToken cancellationToken = default) { throw null; }
+#else
```
I've removed this method from the implementation and the ref assemblies. At the moment, we get the metadata for a result set by getting the `SqlDataReader.MetaData` property, which synchronously waits on the network. A future PR can perform the work needed to unpick the async-over-sync here and provide a fully async implementation.

paulmedynski commented:
/azp run

edwardneal committed:
Merge main

edwardneal commented:
I've updated this branch from after the SqlConnection merges were applied, could I have a CI run please?

paulmedynski commented:
/azp run

edwardneal committed:
Merge branch 'main' into issue-646

edwardneal committed:
Merge main

edwardneal commented:
Yes - merged, thanks.

edwardneal commented:
Updating to hold this open for now.

edwardneal committed:
Merge main

edwardneal commented:
This PR is not stale.

mdaigle commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlMetaDataFactory.cs:
```diff
@@ -159,13 +161,20 @@ private DataRow FindMetaDataCollectionRow(string collectionName)
+        public async Task<DataTable> GetSchemaAsync(DbConnection connection, string collectionName, string[] restrictions, CancellationToken cancellationToken) =>
+            await GetSchemaCore(connection, collectionName, restrictions, isAsync: true, cancellationToken).ConfigureAwait(false);
+
+        public async ValueTask<DataTable> GetSchemaCore(DbConnection connection, string collectionName, string[] restrictions, bool isAsync, CancellationToken cancellationToken)
```
This can be made sync if you have the switch cases directly return their ValueTasks.

edwardneal committed:
Merge main

edwardneal committed:
Update ref assemblies post-merge

paulmedynski commented:
@edwardneal - Can you rebase/merge from main?  The macOS failure is strange, and I know @benrr101 has made many build changes lately that are passing elsewhere.

edwardneal committed:
Merge branch 'main' into issue-646

edwardneal commented:
I've just merged - hopefully that fixes the failure.

paulmedynski commented:
/azp run
